Monitor deployment playbook

rzadp opened the pull request:
- Introduces playbook for deploying monitor very similarly as #126 
    - Created #134 to add remote logging in subsequent PR
- Closes #10
- I'm not sure what values `MONITOR_LEFT_TX_THRESHOLD`, `MONITOR_HOME_DEPLOYMENT_BLOCK` and others should be put in `dai.yml` and `wetc.yml`. I took defaults from `monitor/.env.example`
- I've tested the playbook using molecule and digitalocean droplet using `dai.yml`

rzadp commented:
I have added #136 to add cron schedule separately as it's home-time but still I wanted to roll out PR before the weekend. @phahulin @ArseniiPetrovich Please confirm that we do want to have a step in the playbook to add a cron entry?

phahulin commented:
Yes, cron is the way we currently use monitors

phahulin commented on deployment/roles/monitor/tasks/jumpbox.yml:
```diff
@@ -0,0 +1,12 @@
+  args:
+    chdir: "{{ bridge_path }}/monitor"
+- name: Initial checks 1/2
+  shell: docker-compose exec -T monitor /bin/bash -c "node checkWorker.js"
```
Let's add timeout here: 
```
shell: docker-compose exec -T monitor /bin/bash -c "timeout -s 9 5m node checkWorker.js"
```

rzadp replied:
Done.

phahulin commented on deployment/roles/monitor/tasks/jumpbox.yml:
```diff
@@ -0,0 +1,12 @@
+  args:
+    chdir: "{{ bridge_path }}/monitor"
+- name: Initial checks 2/2
+  shell: docker-compose exec -T monitor /bin/bash -c "node checkWorker2.js"
```
Also add timeout here

rzadp replied:
Done.

phahulin commented on deployment/group_vars/dai.yml:
```diff
@@ -39,3 +39,12 @@ UI_HOME_EXPLORER_ADDRESS_TEMPLATE: https://blockscout.com/poa/dai/address/%s
+MONITOR_GAS_LIMIT: 300000
+MONITOR_GAS_PRICE_FALLBACK: 21
+MONITOR_LEFT_TX_THRESHOLD: 100
+MONITOR_CRON_SCHEDULE: "* * * * *"
```
Currently we have separate schedules for checkWorker and ckeckWorker2: `0-59/3 * * * *` for the first one and  `1-59/3 * * * *` for the second one, so that at least for the first minute they don't overlap. This applies to all bridges.

rzadp replied:
This would be done in #138 (Currently this schedule is not used)

phahulin commented on deployment/group_vars/dai.yml:
```diff
@@ -39,3 +39,12 @@ UI_HOME_EXPLORER_ADDRESS_TEMPLATE: https://blockscout.com/poa/dai/address/%s
+
+## Monitor
+MONITOR_PORT: 3003
+MONITOR_HOME_DEPLOYMENT_BLOCK: 0
```
Here it's `759`

rzadp replied:
Done.

phahulin commented on deployment/group_vars/dai.yml:
```diff
@@ -39,3 +39,12 @@ UI_HOME_EXPLORER_ADDRESS_TEMPLATE: https://blockscout.com/poa/dai/address/%s
+## Monitor
+MONITOR_PORT: 3003
+MONITOR_HOME_DEPLOYMENT_BLOCK: 0
+MONITOR_FOREIGN_DEPLOYMENT_BLOCK: 0
```
And this is `6478417`

rzadp replied:
Done.

phahulin commented on deployment/group_vars/wetc.yml:
```diff
@@ -39,3 +39,12 @@ UI_HOME_EXPLORER_ADDRESS_TEMPLATE: https://blockscout.com/etc/mainnet/address/%s
+
+## Monitor
+MONITOR_PORT: 3001
+MONITOR_HOME_DEPLOYMENT_BLOCK: 0
```
This value should be `7703292`

rzadp replied:
Done.

phahulin commented on deployment/group_vars/wetc.yml:
```diff
@@ -39,3 +39,12 @@ UI_HOME_EXPLORER_ADDRESS_TEMPLATE: https://blockscout.com/etc/mainnet/address/%s
+## Monitor
+MONITOR_PORT: 3001
+MONITOR_HOME_DEPLOYMENT_BLOCK: 0
+MONITOR_FOREIGN_DEPLOYMENT_BLOCK: 0
```
And `7412459` here

rzadp replied:
Done.

akolotov commented:
@phahulin 
> As a suggestion for discussion, would it be difficult to have this playbook support multiple monitorings for multiple networks running on the same host?

It is a good suggestion. I think we definitely need to create a separate issue for this. Especially since we need to find a better approach to do so. That's why I created https://github.com/poanetwork/tokenbridge/issues/138.

akolotov commented on deployment/roles/monitor/templates/tokenbridge-monitor.j2:
```diff
@@ -0,0 +1,76 @@
+# Description:       Enable service provided by daemon.
+### END INIT INFO
+
+WORKDIR="{{ '/home/' + compose_service_user | default('poadocker') + '/' + bridge_path + '/ui' if bridge_path[:1] != "/" else bridge_path + '/ui' }}"
```
is `/ui` correct directory here?

rzadp replied:
Updated, sorry for sloppiness.

akolotov commented on deployment/roles/monitor/tasks/servinstall.yml:
```diff
@@ -0,0 +1,17 @@
+# This role creates a tokenbridge-monitor service which is designed to manage docker-compose ui deployment.
```
replace `ui` with `monitor`

rzadp replied:
Done.

akolotov commented on deployment/roles/monitor/tasks/jumpbox.yml:
```diff
@@ -0,0 +1,12 @@
+- name: Launch container
```
what exactly is this tasks doing? When is it going to be run? Once during deployment?  Periodically by cron? Could you clarify?

rzadp replied:
1. The `Launch container` launches the monitor server, serving data over HTTP GET requests. It is started during deployment, and setup to start after reboot and available for administration as a `tokenbridge-monitor` service - much like `poabridge` and `tokenbridge-ui` services.
2. The data, served by the above server, is produced by running `checkWorker.js` and `checkWorker2.js`
    - I run it initially in `Initial checks ...` steps because otherwise, the monitor server breaks when there's no data to serve
    - These scripts should be run periodically by cron - I created #136 and hope to do it in separate PR

akolotov replied:
thanks. Is my understanding correct and the monitor server cannot start without the data produces by workers? Does it mean that the proper way to fix it is to run workers before starting the server. Imagine the situation when we restart `tokenbridge-monitor`, it will reset the file system and no data will exist for the monitor server, so the restart will be unsuccessful.

So, could you consider to remove `Initial checks 1/2` and `Initial checks 2/2` but add running of the workers in the start script of `package.json`?
https://github.com/poanetwork/tokenbridge/blob/f4151230ca9c678932d6ca0a0737ada97b69346d/monitor/package.json#L8

rzadp replied:
You're spot on.
I'll do the changes.

akolotov commented on deployment/CONFIGURATION.md:
```diff
@@ -40,9 +40,13 @@ cp hosts.yml.example hosts.yml
           ansible_user: <user>
         <host_ip_C>:
           ansible_user: <user>
+    monitor:
```
Should you also update `EXECUTION.md`?

rzadp replied:
Updated.

rzadp commented:
@akolotov @phahulin Thanks for your comments. Please review again after my updates.

The outstanding issues I wanted to leave out of the scope of this PR are #134 #136 #138

phahulin commented on deployment/group_vars/dai.yml:
```diff
@@ -39,3 +39,12 @@ UI_HOME_EXPLORER_ADDRESS_TEMPLATE: https://blockscout.com/poa/dai/address/%s
 UI_FOREIGN_GAS_PRICE_FALLBACK: 10000000000
+
+## Monitor
+MONITOR_PORT: 3003
```
Monitor port here is different from the rest of networks where it's 3001

rzadp replied:
Good catch. Actually I wanted monitor ports for all bridges to be 3003, different than UI's 3001. Updated now.

phahulin commented on deployment/roles/monitor/tasks/jumpbox.yml:
```diff
@@ -0,0 +1,12 @@
+  args:
+    chdir: "{{ bridge_path }}/monitor"
+- name: Initial checks 1/2
+  shell: docker-compose exec -T monitor /bin/bash -c "timeout -s 9 5m node checkWorker.js"
```
Should this command (and the one below) be run once during initial launch? It doesn't look like they worked in my test because no `responses/*.json` files were created in container's filesystem
```
~$ sudo docker exec -it monitor_monitor_1 /bin/bash
root@0c883ae1a263:/mono/monitor# ls
alerts.js	checkWorker2.js  crontab.example  getBalances.js	 index.js   package.json  stuckTransfers.js  utils
checkWorker.js	checkWorker3.js  eventsStats.js   getShortEventStats.js  logger.js  responses	  test-srv.js	     validators.js
root@0c883ae1a263:/mono/monitor# cd responses/
root@0c883ae1a263:/mono/monitor/responses# ls
# empty list
```

phahulin replied:
And later if I run workers manually, these files are created
```
root@0c883ae1a263:/mono/monitor# node checkWorker
checkWorker.js   checkWorker2.js  checkWorker3.js  
root@0c883ae1a263:/mono/monitor# node checkWorker.js 
2019-07-08T19:57:09.441Z (+0ms) [validators] Fetched gasprice: 1
root@0c883ae1a263:/mono/monitor# node checkWorker2.js 
root@0c883ae1a263:/mono/monitor# ls responses/
alerts.json  eventsStats.json  getBalances.json  validators.json
```

rzadp commented:
@akolotov @phahulin I have addressed your comments which are both related to running the `checkWorker` scripts. https://github.com/poanetwork/tokenbridge/pull/133#discussion_r301274852 https://github.com/poanetwork/tokenbridge/pull/133#discussion_r301151250

I have tested with `dai` config that the `responses/xxx` files are created after running playbook, and also that it works after restarting the `tokenbridge-monitor` service.

Note: Since there are changes not only to the playbook but also to `monitor`, when testing the playbook please make sure to change `bridge_repo_branch: monitor-playbook`
